[tensor-widget] Add slicing control to support 3D+ tensors

caisq opened the pull request:
* Motivation for features / changes
  * Continue developing TensorWidget

* Technical description of changes
  * Add a slicing control UI, included in the newly created file silcing-control.ts
  * This UI allows the user to  move between slices and to flexibly change which dimensions of a 3D+ tensor are sliced and which are viewed in the value table.
  * The dimension selection is supported by an ad hoc dropdown menu.

* Screenshots of UI changes
![image](https://user-images.githubusercontent.com/16824702/63184491-43bb4e00-c00c-11e9-879f-28d817415018.png)

The dropdown menu for swapping a viewing dimension with a slicing one:
![image](https://user-images.githubusercontent.com/16824702/63186411-30f74800-c011-11e9-86d0-e1bc9e856aef.png)

The div for a slicing dimension becomes a number input box when clicked.
![image](https://user-images.githubusercontent.com/16824702/63186441-44a2ae80-c011-11e9-9950-47e9909ccc90.png)

- Co-author: @bileschi

bileschi commented on tensorboard/components/tensor_widget/slicing-control.ts:
```diff
@@ -0,0 +1,335 @@
+   */
+  private renderDropdownMenuItems(
+    dropdown: HTMLDivElement,
+    top: number,
```
consider using a coord.XY pattern here, to collect the two values with similar meaning & use into a common object. 

/**
 * Represents a position in a planar coordinate frame.
 */
export interface XYCoordinates {
  x: number;
  y: number;
}

caisq replied:
The `top` and `left` arg names correspond nicely to the return type of [getBoundingClientRect()](https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect). I'd argue that creating a Coordinate-like interface or struct only adds confusion, because coordinates are usually x and y, but in the HTML coordinate system, things are called left, right, top and bottom. It would be confusing if force associate left with x and top with y; it'll also be confusing if you let the members of the interface be called `left` and `top` and so forth.

caisq commented on tensorboard/components/tensor_widget/shape-utils.ts:
```diff
@@ -80,3 +80,35 @@ export function getDefaultSlicingSpec(shape: Shape): TensorViewSlicingSpec {
+ * @return `true` if and only if the slicing dimensions and/or the viewing
+ *   dimensions differ between the `spec0` and `spec1`.
+ */
+export function dimensionsDiffer(
```
I renamed to to `areSlicingSpecsCompatible` for better clarity.

caisq commented on tensorboard/components/tensor_widget/shape-utils.ts:
```diff
@@ -80,3 +80,35 @@ export function getDefaultSlicingSpec(shape: Shape): TensorViewSlicingSpec {
+  spec0: TensorViewSlicingSpec,
+  spec1: TensorViewSlicingSpec
+): boolean {
+  if (spec0.viewingDims[0] !== spec1.viewingDims[0]) {
```
If the length is less than 1, then `[0]` will give you `undefined`, which works for the logic here.

caisq commented:
@bileschi Thanks for the very helpful review. Listed you as a co-author in this PR if that's okay with you.

- Regarding the point of the dropdown menu hard to get rid of, realize that the dropdown menu has a 'mouseleave' listener that closes itself if the mouse cursor enters and the leaves it, even without clicking. But to address the problem you pointed out, I added an extra 'mouseleave' listener, i.e., to the slicing control bar, such that if the mouse goes out of that region, the dropdown menu goes away by itself.

- Regarding the point that it's hard to keep track of which dimensions are being swapped in the dropdown menu, I added hover-based visual highlighting for the dropdown menu. For example, see the screenshot below. When you hover the mouse cursor on a menu item of the dropdown, the dimension with which it'll be swapped with will be highlighted by a color outline and bold font weight.

![image](https://user-images.githubusercontent.com/16824702/63313625-194ee680-c2d3-11e9-9534-0eca9633e8fc.png)

caisq commented on tensorboard/components/tensor_widget/slicing-control.ts:
```diff
@@ -0,0 +1,335 @@
+  private slicingSpec: TensorViewSlicingSpec;
+
+  // Constituent UI components.
+  // private rootDiv: HTMLDivElement;
```
Removed. Thanks for the catch.

caisq commented on tensorboard/components/tensor_widget/slicing-control.ts:
```diff
@@ -0,0 +1,335 @@
+    private readonly shape: Shape,
+    private readonly onSlicngSpecChange?: OnSlicingSpecChangeCallback
+  ) {
+    this.rank = this.shape.length;
```
This goes back to how the tensor widget works. If the underlying tensor's rank changes, the UI widget won't reflect that until its `render()` method is called again.  However, when the `render()` method is called, it'll always create a new SlicingControl instance, which will reflect the up-to-date rank.

caisq commented on tensorboard/components/tensor_widget/slicing-control.ts:
```diff
@@ -0,0 +1,335 @@
+   */
+  render(slicingSpec?: TensorViewSlicingSpec) {
+    if (slicingSpec != null) {
+      this.slicingSpec = JSON.parse(JSON.stringify(slicingSpec));
```
This is a way to deep-copy the object. See the accepted answer in https://stackoverflow.com/questions/122102/what-is-the-most-efficient-way-to-deep-clone-an-object-in-javascript

We want to avoid mutating the input argument `slicingSpec`.

wchargin replied:
Sure, but note jeresig’s comment and the actual benchmarks showing that
stringify–parse is (as should be expected) the _slowest_ way to copy:
<https://stackoverflow.com/a/5344074/>

We already have a lodash dependency; could you please use `_.cloneDeep`
instead?

caisq replied:
The object being deep-copied here is tiny. This method is not called frequently. If the performance becomes and issue in the future, we can consider using lodash. But for now, I think the benefit of one fewer dependency outweighs the tiny performance gain. (TensorWidget is intended to be usable as a standalone package)

bileschi commented on tensorboard/components/tensor_widget/shape-utils.ts:
```diff
@@ -80,3 +80,39 @@ export function getDefaultSlicingSpec(shape: Shape): TensorViewSlicingSpec {
+ * Check if two slicing specs involve the compatible dimension arrangements.
+ *
+ * "Compatible dimension arrangement" means the same dimensions are used
+ * for viewing and the same dimensions are used for slcing.
```
nit: sl*i*cing

caisq replied:
Fixed.

caisq commented:
Regarding the point that there is some extra space for tensor name, note that the top-right part of the widget is reserved for health pill, which hasn't been implemented yet, but will be soon.

caisq commented on tensorboard/components/tensor_widget/shape-utils-test.ts:
```diff
@@ -116,3 +121,107 @@ describe('getDefaultSlicingSpec', () => {
+    expect(dimensionsDiffer(spec0, spec1)).to.be.false;
+  });
+
+  it('Differen slicing and viewing dimensions are captured', () => {
```
Fixed.

caisq commented:
Added commas to the slicing control bar,  so that it looks like a Python/numpy slicing string, which IMO facilitates understanding of the UI. 

Screenshot:
![image](https://user-images.githubusercontent.com/16824702/63314810-e1e23900-c2d6-11e9-9cb4-e1f996cdebef.png)

wchargin commented on tensorboard/components/tensor_widget/slicing-control.ts:
```diff
@@ -0,0 +1,335 @@
+              Math.floor(dimSize) != dimSize
+            ) {
+              // Reject invalid value.
+              dimInput.value = `${this.slicingSpec.slicingDimsAndIndices[slicingDims.indexOf(i)].index}`;
```
Please use `String(foo)` instead of `` `${foo}` ``.

caisq replied:
Done throughout this file.

wchargin commented on tensorboard/components/tensor_widget/slicing-control.ts:
```diff
@@ -0,0 +1,335 @@
+              slicingDims.indexOf(i)
+            ].index = newIndex;
+            dimControl.textContent = `${newIndex}/${dimSize}`;
+            this.onSlicngSpecChange(this.slicingSpec);
```
s/Slicng/Slicing/?

caisq replied:
Fixed. Thanks for the catch.
